config.dev.yaml

```diff
@@ -0,0 +1,35 @@
+issuer: http://127.0.0.1:5556/dex
+
+storage:
+  type: memory
+  config:
+    file: var/sqlite/dex.db
```

Comment:
this path doesn't correspond to `/var/dex` from #1958

Reply:
This config is for local development, relative to the project directory.

Reply:
The memory storage backend doesn't have any config options, so it is confusing to keep sqlite3-specific settings here.

So I recommend removing `config` completely if the `storage.type` is set to `memory`.

Reply:
Damn, I missed that all your comments were related to the `config.dev.yaml` file. 😄 

My original intention was to use the memory storage in the default config and sqlite in the dev config. I'll fix that in a bit.

Reply:
No problem, but keep in mind that directory `var/sqlite/` doesn't exist, so the Dex will fail to start if you use this config file (both locally and in docker). That was my original proposal to put `/var/dex/dex.db` here so it would work in docker, at least.
Maybe other developers could express their opinion if this change is acceptable? I really don't want to break their workflow...

If your intention with this file was to allow instantly running dex Docker image, you should put `file: /var/dex/dex.db` here.
But if you prefer running dex locally, then having something like `file: dex.db` is better.

Or, a more hackish approach: add empty file `var/dex/.keep` to git and use `file: var/dex/dex.db`. It will run both in docker and locally, provided you start local dex from the git repo root directory. :wink:

Reply:
This file is intended for local development, so Dex will most probably run locally (that's how we develop Dex today). So for this file, I don't really care much about making it work in docker.

As for the non-existent path: I haven't really thought about it. Ideally, this file would be copied to `config.yaml` (so that developers can change it as needed) by some `make up` or similar command, which could also create the var directory?

Or perhaps I should just revert it to `dex.db`? I don't know what others think, but it physically hurts me to see random files hanging around in the project root.

Reply:
Hey @mouchar @sagikazarmark I cam e across this while working on a different issue.
I'm currently deploying dexidp in docker-compose. I am using the  manifests in this issue.
https://github.com/traefik/traefik/issues/7814
Let me know if this could enhance dev experience for running dexidp with docker.

Reply:
@jawabuu Thanks! I'm not entirely sure what I should be looking at though. We are in the middle of improving our dev environment, so things will change anyway.

Reply:
From a developer experience perspective, the major drawback of Dex in the Docker deployment I can see is the need for an external config file. For quick tests and prototypes, running `docker run dex` is simply not enough, you need to get an example config file from the git repo, or write your own from scratch. 12-factor app methodology recommends configuring apps using the environment. Environment cfg is currently very limited (according to `git grep os.Getenv`). Having the possibility to run the container straight from an image (without the need to download and bind-mount a config file) would be great, especially for the docker/docker-compose use case.
On the other hand, running Dex in K8S with external config is fine - proper tools already exist in the K8S ecosystem that will help you to generate a Dex config file in advance.

Just my 2¢ based on my experience with Dex (local run, Docker, Docker Compose, and Kubernetes+Helm).

Reply:
@mouchar I agree, configuring Dex is a bit cumbersome today.

#1965 should help with using environment variables. There are also flags now for a few options (where it makes sense): addresses mainly (http, https, grpc, telemetry).

This PR should also improve the default configuration so that you can run Dex without a config file, albeit it might not be useful for anything.

Reply:
Hey @mouchar @sagikazarmark 
Thank you for getting back to me.

In my opinion having a docker-compose file & a default config file in the repo that can consistently & reliably bring up dexidp and quickly demonstrate it's features provides an unmatched developer experience.
That is, as long as you have docker just clone the example & run
`docker-compose up -d`
that's it 🙂
As this is a dev environment running locally, bind mounts are expected and actually preferred to allow one to play around with the different config options.
Additionally, the docker-compose should work hand in hand with existing examples.
You could have the following structure,
`examples/local`
`examples/docker-compose`
`examples/[anything]`
to give you flexibility in future.
The `docker-compose` directory will have at least the following files
`examples/docker-compose/docker-compose.yml`
`examples/docker-compose/config.yml`
Let me post the files to demonstrate

Reply:
#### Structure
```
+---compose
|       dex-config.yml
|       docker-compose.yml
```
####  dex-config.yml
```dex-config.yml
issuer: http://localhost:5556
storage:
  type: memory
web:
  http: 0.0.0.0:5556
telemetry:
  http: 0.0.0.0:5558
logger:
  level: "debug"
staticClients:
- name: 'Example'
  id: example-app
  secret: ZXhhbXBsZS1hcHAtc2VjcmV0
  redirectURIs:
  - 'http://localhost:6555/callback'
  - 'http://dex-app:6555/callback'
connectors:
- type: mockCallback
  id: mock
  name: Example
enablePasswordDB: true
staticPasswords:
- email: "admin@example.org"
  # bcrypt hash of the string "password"
  hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
  username: "admin"
  userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
```
####  docker-compose.yml
``` docker-compose.yml
version: '2.1'
networks:
  dex:
services:  
  dex:
    image: ghcr.io/dexidp/dex:v2.27.0
    healthcheck:
      test: ["CMD", "wget", "-qO-", "-t1", "http://localhost:5556/healthz"]
      interval: 30s
      timeout: 60s
      retries: 5
    command: ["serve", "/etc/config.yml"]
    ports:
      - 5556:5556
    volumes:
      - "./dex-config.yml:/etc/config.yml:ro"
    networks:
      dex:
        aliases:
          - localhost
  
  dex-app:
    image: obitech/dex-example-app
    command: ["--debug","--listen=http://dex-app:6555","--redirect-uri=http://localhost:6555/callback","--issuer=http://localhost:5556"]
    ports:
      - 6555:6555
    networks:
      - dex
    depends_on:
      dex:
        condition: service_healthy

```

Above is the most basic example that you can test out.

Reply:
@mouchar @sagikazarmark 
If acceptable, you may have a more advanced setup that includes e.g. a mysql or postgres service and an example `.env` file to demonstrate env variable features.
This is simple enough to do.
You should probably also host the `dex-example-app` image yourselves.